Modules/itertoolsmodule.c

```diff
@@ -2570,6 +2570,7 @@ typedef struct {
     PyObject *pool;         /* input converted to a tuple */
     Py_ssize_t *indices;    /* one index per result element */
     PyObject *result;       /* most recently returned result tuple */
+    Py_ssize_t n_r_delta;   /* maximum possible value for first index (n - r) */
```

Comment:
Does adding the `n_r_delta` member buy something of value? Computing `n-r` when needed  is very cheap, should overlap with memory reads anyway, and should be easy for an optimizing compiler to recognize as an invariant subexpression.

Reply:
It is a precomputed value used in a few places which can loop (including next() itself), so whether re-computation is actually optimized or not by the compiler is an open question, but I doubt it.

When called millions of times in a row, which is what I understand what combinations can end up to, the "cheap" re-computation at every step can overtake the small penalty of an extra combinations state variable.

Reply:
It won't be reused _across_ calls, no. The only way to know is to time it. If it actually does help, then instead of adding a new member to the struct, I'm sure it would do just as well to add, e.g.,

```c
    const Py_ssize_t n_r_delta = n - r;
```

near the top of the function. That tells the compiler directly that it's a function-invariant value, even in a debug build, and without putting additional pressure on caches.

Reply:
And noting that computing `n-r` at all is something of a red herring. Regardless of `r`, the largest value at the highest index in the tuple is `n-1`. So instead of bothering with `n-r`, a variable (say `hi`) could be initialized to `n-1`, and decremented by 1 each time through the search loop. I believe that would also be more obvious.

Reply:
The benchmarks and this PR were done in January, so whether your statements are correct or not can be proven by running new benchmarks. What I did then was to move any "constant" values off the processing logic. This was my first Python PR so if you feel that using a "const" has the identical effect as the intent of this change, then by all means pursue it. I just thought the co object was the best place to put such an invariable value. I also found that small changes to how many local variables a low-level loop uses can affect the resulting assembly code (and hence the speed) in a lot of ways. YMMV.

Reply:
I apologize on behalf of the core devs that this sat unaddressed for so long. When Raymond added me, I leaped right on it.

But, for good or ill,, the burden remains on you to "sell" it. "YMMV" is worse than you may think :wink: - for speedups so small, they vary across platforms (including HW, compiler, compiler release, and compiler options) too. 

BTW, under Microsoft's Visual Studio 2022 in a release build, I got similar speedups _just_ by introducing a `const` variable for `n-r` in the current implementation, and replacing the single instance of `n-r` with it. So that specific platform is doing a poor job at optimizing `n-r`.

It's a minor speedup at best, so I'm happy to put _some_ effort into moving this along, but not a lot of effort. If you're worn out from putting effort into it too, that's understandable, but I'll have to close the PR. If you want to pursue this, you'll need to address comments made.